qiskit/synthesis/unitary/qsd.py

```diff
@@ -63,21 +65,20 @@ def qs_decomposition(
 
     This decomposition is described in arXiv:quant-ph/0406176.
 
-    Arguments:
-       mat (ndarray): unitary matrix to decompose
-       opt_a1 (bool): whether to try optimization A.1 from Shende. This should eliminate 1 cnot
-          per call. If True CZ gates are left in the output. If desired these can be further decomposed
-          to CX.
-       opt_a2 (bool): whether to try optimization A.2 from Shende. This decomposes two qubit
-          unitaries into a diagonal gate and a two cx unitary and reduces overal cx count by
-          4^(n-2) - 1.
-       decomposer_1q (None or Object): optional 1Q decomposer. If None, uses
-          :class:`~qiskit.quantum_info.synthesis.one_qubit_decomposer.OneQubitEulerDecomser`
-       decomposer_2q (None or Object): optional 2Q decomposer. If None, uses
-          :class:`~qiskit.quantum_info.synthesis.two_qubit_decomposer.two_qubit_cnot_decompose
+    Args:
```

Comment:
It seems this function has not been appearing in the API docs before, so it would be good to take the chance and fix these docs. Do we know what type the `decomposer_1q` and `decomposer_2q` are? If yes, we could add them as typehints in the function signature and remove all typehints from the docstring. If we have them in both places, I think the ones in the signature are overridden 🤔

Reply:
Regarding the typehints: From how they are used it looks like they both are either `None`, or a function that takes a matrix as input and returns a decomposition, i.e. 
```python
Callable[[np.ndarray], QuantumCircuit] | None 
```

Regarding the documentation: I was referring to the API doc of `qs_decomposition`. It seems the linked comment is concerned with making the `two_qubit_cnot_decompose` available upon importing Qiskit, not whether to make it API documented, does it? 🤔

Reply:
this function is still available via importing, see line 156 of [`qiskit/synthesis/__init__.py`](https://github.com/Qiskit/qiskit/pull/11460/files#diff-c69f61f5e44a7d3a64491177cfb43f586cacebd380d521658f64967d49baafb9):
https://github.com/Qiskit/qiskit/blob/13e63a3fe040b9ff1a6e1b6150034b2881d39b53/qiskit/synthesis/__init__.py#L156

Reply:
Both `decomposer_1q` and `decomposer_2q` take class instances. Perhaps it would be more transparent if they use the classes directly. I think at the time I thought `two_qubit_cnot_decompose` was a higher level interface or avoided circular import but it sounds like it is actually for backwards compatibility.

So perhaps change line 104 to,
`decompser_2q = two_qubit_decompose.TwoQubitBasisDecomposer(CXGate())`

Reply:
OK, I also removed `two_qubit_cnot_decompose` from `qiskit.synthesis.__init__`